src/cmd/linuxkit/moby/build.go

```diff
@@ -427,6 +453,53 @@ func tarAppend(iw *tar.Writer, tr *tar.Reader) error {
 	return nil
 }
 
+// Attempt to decompress a Linux kernel image
+// arm64 kernel images are plain gzip
+// x86 kernel images are in bzImage format, where the offset for the gzip'ed image is described in the header
+// s390x kernels images are not compressed.
+func decompressKernel(src *bytes.Buffer) (*bytes.Buffer, error) {
+	const gzipMagic = "\037\213\010" // Note, this is for gzip with max decoding used by the kernel
```

Comment:
Perhaps rather than this, which encodes a specific set of options you could just try and then handle `gzip.ErrHeader`. i.e. let the gzip library probe the header.